tools/fuzzers/fuzzilli/fuzzilli.cpp

```diff
@@ -0,0 +1,214 @@
+#include <cerrno>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+#define REPRL_CRFD 100
+#define REPRL_CWFD 101
+#define REPRL_DRFD 102
+#define REPRL_DWFD 103
+
+#define SHM_SIZE 0x100000
+#define MAX_EDGES ((SHM_SIZE - 4) * 8)
+
+#define CHECK(cond)                            \
+  if (!(cond)) {                               \
+    fprintf(stderr, "\"" #cond "\" failed\n"); \
+    _exit(-1);                                 \
+  }
+
+struct SharedMemData {
+  uint32_t numEdges;
+  unsigned char edges[];
+};
+
+struct SharedMemData *shmem;
+
+uint32_t *__edges_start, *__edges_stop;
+void __sanitizer_cov_reset_edgeguards() {
+  uint64_t N = 0;
+  for (uint32_t *x = __edges_start; x < __edges_stop && N < MAX_EDGES; x++)
+    *x = ++N;
+}
+
+extern "C" void __sanitizer_cov_trace_pc_guard_init(
+    uint32_t *start,
+    uint32_t *stop) {
+  // Avoid duplicate initialization
+  if (start == stop || *start)
+    return;
+
+  if (__edges_start != nullptr || __edges_stop != nullptr) {
+    fprintf(
+        stderr,
+        "Coverage instrumentation is only supported for a single module\n");
+    _exit(-1);
+  }
+
+  __edges_start = start;
+  __edges_stop = stop;
+
+  // Map the shared memory region
+  const char *shm_key = getenv("SHM_ID");
+  if (!shm_key) {
+    puts("[COV] no shared memory bitmap available, skipping");
+    shmem = (struct SharedMemData *)malloc(SHM_SIZE);
+  } else {
+    int fd = shm_open(shm_key, O_RDWR, S_IREAD | S_IWRITE);
+    if (fd <= -1) {
+      fprintf(
+          stderr, "Failed to open shared memory region: %s\n", strerror(errno));
+      _exit(-1);
+    }
+
+    shmem = (struct SharedMemData *)mmap(
+        nullptr, SHM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (shmem == MAP_FAILED) {
+      fprintf(stderr, "Failed to mmap shared memory region\n");
+      _exit(-1);
+    }
+  }
+
+  __sanitizer_cov_reset_edgeguards();
+
+  shmem->numEdges = stop - start;
+  printf(
+      "[COV] edge counters initialized. Shared memory: %s with %u edges\n",
+      shm_key,
+      shmem->numEdges);
+}
+
+extern "C" void __sanitizer_cov_trace_pc_guard(uint32_t *guard) {
+  // There's a small race condition here: if this function executes in two
+  // threads for the same edge at the same time, the first thread might disable
+  // the edge (by setting the guard to zero) before the second thread fetches
+  // the guard value (and thus the index). However, our instrumentation ignores
+  // the first edge (see libcoverage.c) and so the race is unproblematic.
+  uint32_t index = *guard;
+  // If this function is called before coverage instrumentation is properly
+  // initialized we want to return early.
+  if (!index)
+    return;
+  shmem->edges[index / 8] |= 1 << (index % 8);
+  *guard = 0;
+}
+
+//
+// END FUZZING CODE
+//
+
+
+int main(int argc, char **argv) {
+  if (argc != 2 || strcmp(argv[1], "--replr") != 0)
+    return 0;
+  // replr mode
+  auto runtime = makeHermesRuntime();
```

Comment:
I'm not familiar with what a "REPLR" server means, but if the intention is to test each script separately, it would be better to instantiate the runtime once per script.
If you share a runtime amongst multiple scripts, then one script may only fail because a previous script ran in the same process.

Reply:
REPLR is meant to use one runtime for a lot of scripts to improve efficiency. I checked V8 and JSC's impl and it seems that a common way to do separation is by 
```js
function x(){ [PAYLOAD] }; x();
```
It is updated now.

Reply:
Anything that could reasonably be corrupted during a run needs to be reset, in each loop of the REPRL (read evaluate print reset loop).

The issue is if an input corrupts some internal state but doesn't crash, a later round will crash and the fuzzer will misattribute it to the later input, without knowing about the previous round. Reseting the state each time ensures a crash is caused solely by that sample, and can be reproduced during triage.

Reply:
I don't think there is a mechanism here that allows resetting. Do you think it is reasonable to just restart runtime each time?

Reply:
I checked QJS and JSC's instrumentations. Both of them do not restart/reset the runtime.

Reply:
I'm looking at jsc.cpp, with the patch applied. The do-while loop is put just above here: https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/jsc.cpp#L3388, and thus creates a new global object each time here: https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/jsc.cpp#L3398

I could be wrong, but that's my understanding.

Reply:
Okay you are right. If there is not any way to clean up the context (@dulinriley is that true?), then we can recreate the runtime every time. I don't see too much overhead so it should be fine...

Reply:
Currently there's no way to clean up the context, so yes recreating the runtime on each step is the right thing to do. Initializing a Hermes runtime takes < 15 ms on an x64 desktop I believe, so the cost should be minimal

Reply:
Done